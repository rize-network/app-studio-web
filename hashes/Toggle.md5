167698f9c571dbbe41332d25be8a5725