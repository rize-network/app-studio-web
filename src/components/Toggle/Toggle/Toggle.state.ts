import { useState } from 'react';
// Declares the useToggleState function which takes defaultToggled parameter to initialize the toggle state.
export const useToggleState = (defaultToggled: boolean) => {
  // Creates a stateful value isHovered for tracking hover state and a function setIsHovered to update that state, initially false.
  const [isHovered, setIsHovered] = useState(false);
  // Creates a stateful value isToggle to store the current toggle state and a function setIsToggled to modify it, initialized with defaultToggled.
  const [isToggle, setIsToggled] = useState(defaultToggled);
  // Returns an object containing the current states and the functions to update them for external use.
  return { isHovered, setIsHovered, isToggle, setIsToggled };
};
