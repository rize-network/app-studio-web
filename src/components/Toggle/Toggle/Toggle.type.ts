export type Shape = 'sharp' | 'rounded' | 'pillShaped';

export type Variant = 'filled' | 'outline' | 'link' | 'ghost';
