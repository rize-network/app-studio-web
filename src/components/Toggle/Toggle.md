# Toggle

Provides interactive UI for binary user input

### **Import**
  ```tsx static
  import { Toggle } from '@app-studio/web';
  ```

### **Default**
```tsx
import React from 'react';
import { Toggle } from '../Toggle';

export const DefaultDemo = () => {
  return <Toggle>Default</Toggle>;
};

```

### **shape**
"Optional 'shape' prop to determine the shape of the toggle."

```tsx
import React from 'react';
import { Horizontal } from '../../Layout/Horizontal/Horizontal';
import { Toggle } from '../Toggle';
import { Shape } from '../Toggle/Toggle.type';
import { View } from 'src/components/Layout/View/View';
export const ShapeDemo = () => (
  <Horizontal gap={15}>
    {['sharp', 'rounded', 'pillShaped'].map((border, index) => (
      <View position="relative" key={index}>
        <Toggle shape={border as Shape}>{border}</Toggle>
      </View>
    ))}
  </Horizontal>
);

```

### **isToggled**
"Optional boolean indicating the current state of the toggle."

```tsx
import React from 'react';
import { Toggle } from '../Toggle';

export const IsToggledDemo = () => {
  return (
    <Toggle isToggled={true} isDisabled>
      IsActive
    </Toggle>
  );
};

```

### **isDisabled**
"Optional boolean to specify if the toggle should be disabled."

```tsx
import React from 'react';
import { Toggle } from '../Toggle';

export const IsDisabledDemo = () => {
  return <Toggle isDisabled>IsDisabled</Toggle>;
};

```

### **variant**
"Optional prop to apply predefined style variants."

```tsx
import React from 'react';
import { Toggle } from '../Toggle';
import { Vertical } from '../../Layout/Vertical/Vertical';
import { Variant } from '../Toggle/Toggle.type';

export const VariantDemo = () => (
  <Vertical gap={15}>
    {['outline', 'link', 'ghost'].map((variant, index) => (
      <Toggle key={index} variant={variant as Variant}>
        {variant}
      </Toggle>
    ))}
  </Vertical>
);

```

### **colorScheme**
"Optional string to customize the color theme of the toggle component."

```tsx
import React from 'react';
import { Toggle } from '../Toggle';
import { Horizontal } from '../../Layout/Horizontal/Horizontal';

export const ColorSchemeDemo = () => (
  <Horizontal gap={10}>
    {[
      'theme.primary',
      'theme.secondary',
      'theme.warning',
      'theme.success',
      'theme.error',
    ].map((color, index) => (
      <Toggle key={index} colorScheme={color}>
        {color}
      </Toggle>
    ))}
  </Horizontal>
);

```

### **onToggle**
"Optional function prop that gets executed when the toggle state changes."

```tsx
import React from 'react';
import { useState } from 'react';
import { Toggle } from '../Toggle';

export const OnToggleDemo = () => {
  const [isToggle, setIsToggled] = useState(false);
  return (
    <Toggle
      onToggle={(state) => {
        setIsToggled(state);
      }}
    >
      {isToggle ? 'On' : 'Off'}
    </Toggle>
  );
};

```

