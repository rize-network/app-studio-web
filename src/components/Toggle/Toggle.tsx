import React, { useEffect } from 'react';
import { ToggleProps } from './Toggle/Toggle.props';
import { useToggleState } from './Toggle/Toggle.state';
import ToggleView from './Toggle/Toggle.view';
// Destructuring properties from ToggleProps to be used within the ToggleComponent.
const ToggleComponent = ({
  children,
  shape,
  colorScheme,
  variant,
  isDisabled,
  // Assigning a default value of false to the isToggled property if none is provided.
  isToggled = false,
  onToggle,
  ...props
}: ToggleProps) => {
  // Initializing toggle state and set state functions using the custom hook useToggleState.
  const { isHovered, setIsHovered, isToggle, setIsToggled } =
    useToggleState(isToggled);
  // Creating an effect that updates the isToggled state whenever the isToggled prop changes.
  useEffect(() => {
    setIsToggled(isToggled);
  }, [isToggled]);
  // Beginning of the JSX block to render the Toggle view component.
  return (
    <ToggleView
      shape={shape}
      colorScheme={colorScheme}
      variant={variant}
      isDisabled={isDisabled}
      isHovered={isHovered}
      setIsHovered={setIsHovered}
      isToggle={isToggle}
      setIsToggled={setIsToggled}
      onToggle={onToggle}
      {...props}
    >
      {children}
    </ToggleView>
  );
};
export const Toggle = ToggleComponent;
