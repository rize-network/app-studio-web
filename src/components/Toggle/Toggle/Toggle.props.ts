import { Shape, Variant } from './Toggle.type';

export interface ToggleProps {
  shape?: Shape;
  isToggled?: boolean;
  isDisabled?: boolean;
  children: React.ReactNode;
  variant?: Variant;
  colorScheme?: string;
  onToggle?: (isToggled: boolean) => void;
  [x: string]: any;
}

export interface ToggleViewProps extends ToggleProps {
  isHovered: boolean;
  setIsHovered: React.Dispatch<React.SetStateAction<boolean>>;
  isToggle: boolean;
  setIsToggled: React.Dispatch<React.SetStateAction<boolean>>;
}
