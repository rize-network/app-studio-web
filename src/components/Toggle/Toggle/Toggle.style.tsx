import { ViewProps } from 'app-studio';
import { Shape, Variant } from './Toggle.type';

export const ToggleShapes: Record<Shape, number | string> = {
  sharp: 0,
  rounded: 4,
  pillShaped: 24,
};

/**
 * Generate toggle variants with proper color combinations based on main color and contrast
 * Similar to Button's getButtonVariants function
 */
export const getToggleVariants = (
  color: string,
  isLight: boolean
): Record<Variant, ViewProps> => ({
  outline: {
    backgroundColor: 'transparent',
    color: color,
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: color,
    _hover: {
      backgroundColor: color,
      color: isLight ? 'color.black' : 'color.white',
      transform: 'translateY(-1px)',
    },
    _active: {
      backgroundColor: color,
      color: isLight ? 'color.black' : 'color.white',
      transform: 'translateY(0)',
    },
    transition: 'all 0.2s ease',
  },
  ghost: {
    backgroundColor: 'transparent',
    color: color,
    borderWidth: 0,
    borderStyle: 'none',
    borderColor: 'transparent',
    _hover: {
      backgroundColor: color,
      color: isLight ? 'color.black' : 'color.white',
      transform: 'translateY(-1px)',
    },
    _active: {
      backgroundColor: color,
      color: isLight ? 'color.black' : 'color.white',
      transform: 'translateY(0)',
    },
    transition: 'all 0.2s ease',
  },
  link: {
    backgroundColor: 'transparent',
    color: color,
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: 'transparent',
    textDecoration: 'underline',
    textUnderlineOffset: '1px',
    textDecorationThickness: '1px',
    _hover: {
      borderColor: color,
      textDecorationThickness: '2px',
      transform: 'translateY(-1px)',
    },
    _active: {
      borderColor: color,
      textDecorationThickness: '2px',
      transform: 'translateY(0)',
    },
    transition: 'all 0.2s ease',
  },
});
