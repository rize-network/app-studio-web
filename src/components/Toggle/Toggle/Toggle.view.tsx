import React, { CSSProperties } from 'react';
import { ToggleViewProps } from './Toggle.props';
import { ToggleShapes } from './Toggle.style';
import { Variant } from './Toggle.type';
import { Center } from 'src/components';

const ToggleView: React.FC<ToggleViewProps> = ({
  children,
  shape = 'rounded',
  colorScheme = 'color.trueGray.400',
  variant = 'ghost',
  isHovered,
  setIsHovered,
  isDisabled,
  isToggle,
  setIsToggled,
  onToggle,
  ...props
}) => {
  const toggleColor = !isDisabled ? colorScheme : 'theme.disabled';

  const isActive = !!(isToggle || isHovered);

  const ToggleVariants: Record<Variant, CSSProperties> = {
    filled: {
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: 'transparent',
    },
    outline: {
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: colorScheme,
    },
    link: {
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: isActive ? toggleColor : 'transparent',
      textDecoration: 'underline',
    },
    ghost: {},
  };

  const handleToggle = () => {
    if (!isDisabled) {
      setIsToggled((prev) => {
        const newState = !prev;
        if (onToggle) {
          // Check if onToggle is defined before calling it
          onToggle(newState);
        }
        return newState; // Expecting the onToggle to possibly do something with the newState
      });
    }
  };

  return (
    <Center
      role="Toggle"
      padding={shape === 'pillShaped' ? '10px 12px' : '8px'}
      width="fit-content"
      color={isActive ? 'color.white' : toggleColor}
      backgroundColor={isActive ? toggleColor : 'transparent'}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
      cursor={isDisabled ? 'not-allowed' : 'pointer'}
      borderRadius={ToggleShapes[shape]}
      onClick={handleToggle}
      {...ToggleVariants[variant]}
      {...props}
    >
      {children}
    </Center>
  );
};

export default ToggleView;
