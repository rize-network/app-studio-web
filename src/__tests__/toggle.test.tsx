// Import necessary libraries and components
import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen } from '@testing-library/react';
import { Shape, Variant } from 'src/components/Toggle/Toggle/Toggle.type';
import { Toggle } from 'src/components/Toggle/Toggle';

// Additional info (mocked for testing if not imported)
export const ToggleShapes: Record<Shape, number | string> = {
  sharp: '0px',
  rounded: '4px',
  pillShaped: '24px',
};

// Test setup and teardown
afterEach(cleanup);

// Test cases
describe('Toggle Component', () => {
  // Basic Rendering
  test('renders Toggle component', () => {
    render(<Toggle>Default</Toggle>);
    const toggleElement = screen.getByRole('Toggle');
    expect(toggleElement).toBeInTheDocument();
  });

  // Shape Variants
  describe('Shape Variants', () => {
    test.each([
      ['sharp', '0px'],
      ['rounded', '4px'],
      ['pillShaped', '24px'],
    ])(
      'applies correct border-radius for shape %s',
      (shape, expectedRadius) => {
        render(<Toggle shape={shape as Shape}>Default</Toggle>);
        const toggleElement = screen.getByRole('Toggle');
        expect(toggleElement).toHaveStyle({ borderRadius: expectedRadius });
      }
    );
  });

  // Variant Styles
  describe('Variant Styles', () => {
    test.each([
      ['outline', { borderStyle: 'solid' }],
      ['link', { textDecoration: 'underline' }],
      ['ghost', {}], // Add expected styles for ghost if necessary
    ])('applies correct styles for variant %s', (variant, expectedStyles) => {
      render(<Toggle variant={variant as Variant}>Default</Toggle>);
      const toggleElement = screen.getByRole('Toggle');
      for (const [key, value] of Object.entries(expectedStyles)) {
        expect(toggleElement).toHaveStyle({ [key]: value });
      }
    });
  });
});

test('Toggle to match snapshot', () => {
  const tree = renderer.create(<Toggle>Default</Toggle>).toJSON();
  expect(tree).toMatchSnapshot();
});
