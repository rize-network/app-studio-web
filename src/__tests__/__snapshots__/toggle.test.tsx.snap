// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle to match snapshot 1`] = `
<div
  className="sc-bdfCDU dmlQCl"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="Toggle"
>
  Default
</div>
`;
